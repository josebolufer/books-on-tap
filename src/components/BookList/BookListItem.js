import { Link as RouterLink } from 'react-router-dom'
import PropTypes from 'prop-types'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'

const useStyles = makeStyles((theme) => ({
  card: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  cardMedia: {
    maxHeight: '250px',
  },
  cardContent: {
    flexGrow: 1,
  },
}))

const BookListItem = ({ book }) => {
  const classes = useStyles()
  const { id, thumbnail, title, price, author } = book

  return (
    <Grid data-testid='book-list-item' item xs={12} sm={6} md={4}>
      <Card raised className={classes.card}>
        <CardMedia
          className={classes.cardMedia}
          src={thumbnail}
          title={title}
          alt={title}
          component='img'
        />
        <CardContent className={classes.cardContent}>
          <Typography
            gutterBottom
            variant='body1'
            component='h2'
            color='primary'
          >
            {title}
          </Typography>
          <Typography
            variant='body2'
            gutterBottom
          >{`By ${author.firstName} ${author.lastName}`}</Typography>
          <Typography variant='body2'>Â£{price.toFixed(2)}</Typography>
        </CardContent>
        <CardActions>
          <Button
            variant='outlined'
            size='small'
            color='primary'
            component={RouterLink}
            to={`/book/${id}`}
          >
            View
          </Button>
        </CardActions>
      </Card>
    </Grid>
  )
}

BookListItem.propTypes = {
  book: PropTypes.object.isRequired,
}

export default BookListItem
