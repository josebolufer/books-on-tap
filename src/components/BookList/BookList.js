import PropTypes from 'prop-types'
import Container from '@material-ui/core/Container'
import { makeStyles } from '@material-ui/core/styles'
import Grid from '@material-ui/core/Grid'
import LinearProgress from '@material-ui/core/LinearProgress'
import Typography from '@material-ui/core/Typography'
import BookListItem from './BookListItem'

const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    '& > * + *': {
      marginTop: theme.spacing(2),
    },
  },
  cardGrid: {
    paddingTop: theme.spacing(8),
    paddingBottom: theme.spacing(8),
  },
}))

const BookList = ({ books, fetching }) => {
  const classes = useStyles()

  if (fetching) {
    return (
      <div data-testid='book-list-fetching' className={classes.root}>
        <LinearProgress color='secondary' />
      </div>
    )
  }

  if (books.length > 0) {
    return (
      <Container className={classes.cardGrid} maxWidth='xl'>
        <Grid container spacing={3}>
          {books.map((book) => (
            <BookListItem key={book.id} book={book} />
          ))}
        </Grid>
      </Container>
    )
  }

  return <Typography variant='body2'>There are 0 books to display.</Typography>
}

BookList.propTypes = {
  books: PropTypes.array,
  fetching: PropTypes.bool,
}

BookList.defaultProps = {
  books: [],
  fetching: false,
}

export default BookList
