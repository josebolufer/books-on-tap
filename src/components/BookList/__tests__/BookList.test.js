import { render, getAllByTestId, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import BookList from '../BookList'

const mockedBooks = [
  {
    author: { firstName: 'Jose', lastName: 'Bolufer' },
    id: '1',
    description: 'The story about a boy that grew up in London',
    isbn: 'asdfghghghghg',
    price: 10,
    stockAmount: 1,
    thumbnail: 'Image 1',
    title: 'A live in London',
  },
  {
    author: { firstName: 'Jose', lastName: 'Bolufer' },
    id: '2',
    description: 'The story about JS',
    isbn: 'asdfghghghghg',
    price: 10,
    stockAmount: 1,
    thumbnail: 'Image 1',
    title: 'A live in Spain',
  },
]

describe('@renders', () => {
  it('in default state', () => {
    const { asFragment } = render(<BookList />, {
      wrapper: MemoryRouter,
    })
    expect(asFragment()).toMatchSnapshot()
  })

  describe('when fetching prop is equal to false', () => {
    it('should render a list of books', () => {
      render(<BookList books={mockedBooks} fetching={false} />, {
        wrapper: MemoryRouter,
      })

      const listItems = getAllByTestId(document.body, 'book-list-item')

      expect(listItems).toHaveLength(2)
    })
  })

  describe('when `fetching` prop is equal to true', () => {
    it('should render a progress bar', () => {
      render(<BookList books={mockedBooks} fetching />, {
        wrapper: MemoryRouter,
      })

      const linearProgressWrapper = getAllByTestId(
        document.body,
        'book-list-fetching'
      )

      expect(linearProgressWrapper).toHaveLength(1)
    })
  })

  describe('when there are no books to show and client is not fetching', () => {
    it('should render text `There are 0 books to display.`', () => {
      render(<BookList />, {
        wrapper: MemoryRouter,
      })

      expect(
        screen.getByText('There are 0 books to display.')
      ).toBeInTheDocument()
    })
  })
})
