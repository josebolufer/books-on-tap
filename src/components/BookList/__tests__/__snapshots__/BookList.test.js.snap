// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@renders in default state 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body2"
  >
    There are 0 books to display.
  </p>
</DocumentFragment>
`;
